clang/include/clang/AST/ExternalASTSource.h

```diff
@@ -326,25 +326,25 @@ struct LazyOffsetPtr {
   ///
   /// If the low bit is clear, a pointer to the AST node. If the low
   /// bit is set, the upper 63 bits are the offset.
-  mutable uint64_t Ptr = 0;
+  mutable uintptr_t Ptr = 0;
 
 public:
   LazyOffsetPtr() = default;
-  explicit LazyOffsetPtr(T *Ptr) : Ptr(reinterpret_cast<uint64_t>(Ptr)) {}
+  explicit LazyOffsetPtr(T *Ptr) : Ptr(reinterpret_cast<uintptr_t>(Ptr)) {}
 
-  explicit LazyOffsetPtr(uint64_t Offset) : Ptr((Offset << 1) | 0x01) {
-    assert((Offset << 1 >> 1) == Offset && "Offsets must require < 63 bits");
+  explicit LazyOffsetPtr(uintptr_t Offset) : Ptr((Offset << 1) | 0x01) {
+    assert((Offset << 1 >> 1) == Offset && "Offsets must fit in addressable bits");
     if (Offset == 0)
       Ptr = 0;
   }
 
   LazyOffsetPtr &operator=(T *Ptr) {
-    this->Ptr = reinterpret_cast<uint64_t>(Ptr);
+    this->Ptr = reinterpret_cast<uintptr_t>(Ptr);
     return *this;
   }
 
-  LazyOffsetPtr &operator=(uint64_t Offset) {
-    assert((Offset << 1 >> 1) == Offset && "Offsets must require < 63 bits");
+  LazyOffsetPtr &operator=(uintptr_t Offset) {
```

Comment:
This may not be correct. The offset is required to be a 64 bit integer while the higher 32 bits are meaningful. Is there any ideas to make it?

Reply:
How can the higher 32 bits be meaningful on a system whose pointers are 32 bits in length?  Put another way: in what situation is a 64-bit offset to a 32-bit pointer valid?

The original message warns about 63 bits because the last bit is needed to tag the value as not-a-pointer (and as an offset).  That doesn't mean values passed here *have* to be 63 bits in length, just that the (naive) implementation that hardcoded 64-bit size for pointers required the offset to fit in N-1 = 63 bits.

Reply:
Well, it's an abstract offset, not a pointer offset.

Reply:
Yeah, we mmap the whole of each module file, so we definitely can't have more than 4GiB of module file data if the host has 32-bit pointers. But `Offset` here is in bits, and with 31 bits of offset, that only covers 256MiB of module data, which is a plausible and reasonable total file size for `.pcm` files even when compiling on a 32-bit host. So I think we do want a 64-bit offset here even on a 32-bit system.

Reply:
If getAddressOfPointer needs to work and there is reluctance to grow the size of the structure then AlignedCharArrayUnion (or equivalent) is probably the way to go (being sure to offset things based on support::endian::system_endianness so the LSBs line up).

Reply:
> uint64_t is not the sufficient size for a 32-bit BE system. How should we proceed? 

I prefer the union based solution here.

Reply:
I think the right solution is to store a `uint64_t` and cast it to a `uintptr_t` before casting back and forth to a pointer type.  I'd like to not hard-code the assumption that AST file sizes can't be out of the range of `uintptr_t` here.  I believe you're right that we do `mmap` the entire AST file right now, but there are other implementation approaches that would support AST files larger than the address space.

Reply:
Some update: the higher 32 bits are required here: https://github.com/llvm/llvm-project/blob/d54953ef472bfd8d4b503aae7682aa76c49f8cc0/clang/include/clang/AST/ExternalASTSource.h#L580-L582

where the `GlobalDeclID`'s higher 32 bits are meaningful. And we may get the `Ptr` converted to `GlobalDeclID` here:

`https://github.com/llvm/llvm-project/blob/d54953ef472bfd8d4b503aae7682aa76c49f8cc0/clang/include/clang/AST/ExternalASTSource.h#L378`

so we need to preserve at least 64 bits in this structure.

Reply:
> I think the right solution is to store a uint64_t and just do an intermediate cast to uintptr_t when going from uint64_t to a pointer type and vice-versa. I'd like to not hard-code the assumption that AST file sizes can't be out of the range of uintptr_t here. I believe you're right that we do mmap the entire AST file right now, but there are other implementation approaches that would support AST files larger than the address space.

This sounds good. When we can convert `Ptr` to a pointer, it must not be an offset, so this converting is safe as long as space is enough.

Reply:
> I think the right solution is to store a `uint64_t` and just do an intermediate cast to `uintptr_t` when going from `uint64_t` to a pointer type and vice-versa. I'd like to not hard-code the assumption that AST file sizes can't be out of the range of `uintptr_t` here. I believe you're right that we do `mmap` the entire AST file right now, but there are other implementation approaches that would support AST files larger than the address space.

That doesn't work for getAddressOfPointer (and doesn't work for CHERI either, but there are many instances of that idiom in Clang to fix...).

Reply:
> so we need to preserve at least 64 bits in this structure.

Hmm.  We're only really preserving 63 bits here, so that might be a problem.

> That doesn't work for getAddressOfPointer

Maybe we can just define away the need for `getAddressOfPointer`.

> (and doesn't work for CHERI either, but there are many instances of that idiom in Clang to fix...).

Is the CHERI problem just the assumption that `uint64_t` is at least as wide as `uintptr_t`, or is there a broader issue with round-tripping through an integer type at all?  At any rate, if we need this type to be implemented substantially differently on CHERI, that's fine, but I'd like to not regress performance on other platforms because of it.

Reply:
> > That doesn't work for getAddressOfPointer
> 
> Maybe we can just define away the need for `getAddressOfPointer`.

If you can do that then the current code works just fine :)



> > so we need to preserve at least 64 bits in this structure.
> 
> Hmm. We're only really preserving 63 bits here, so that might be a problem.
> 
> > That doesn't work for getAddressOfPointer
> 
> Maybe we can just define away the need for `getAddressOfPointer`.
> 
> > (and doesn't work for CHERI either, but there are many instances of that idiom in Clang to fix...).
> 
> Is the CHERI problem just the assumption that `uint64_t` is at least as wide as `uintptr_t`, or is there a broader issue with round-tripping through an integer type at all? At any rate, if we need this type to be implemented substantially differently on CHERI, that's fine, but I'd like to not regress performance on other platforms because of it.

The problem is that uint64_t cannot hold a pointer on CHERI, but uintptr_t can round-trip just fine. The approach I've used downstream in the past when prototyping self-hosted CHERI LLVM was to use uintptr_t when its range is at least uint64_t's and to fall back on uint64_t otherwise (effectively using uint64_t for ILP32 and uintptr_t otherwise). Though if and when I ever get back to cleaning all that up and upstreaming it I'd like to abstract it into a reusable helper rather than replicating the slightly messy std::conditional_t pattern.

Reply:
Okay, so it's just the width.  Why don't we just add a `uintptr64_t` typedef to Compiler.h, with the semantics that it's at least as wide as both `uintptr_t` and `uint64_t`?  Adopting that in places like this seems like it'd be very straightforward, and I'd expect almost all of the violations to be easily identified by `-Wconversion`.

Reply:
That name's reserved for the implementation, and also could be taken to mean "a 64-bit pointer". In CheriBSD we have a `uint64ptr_t` for use in some ioctls, and is exactly the type we're talking about, but you'd clash with it.

Reply:
However, this is getting off topic, because that doesn't help this code for 32-bit PowerPC, unlike some of the other suggestions that also happen to work for CHERI.

Reply:
The 32-bit PPC issue is presumably solved by just adding an intermediate cast.

Reply:
No, it's not. Please re-read the problem description. Casts in the LazyOffsetPtr implementation aren't going to fix anything. The problem is that getAddressOfPointer pretends that the underlying storage is a pointer, when it's really a uint64_t, and on a 32-bit big-endian system `*(void *)&my_uint64 = p` and `my_uint64 = (uint64_t)(uintptr_t)p` store the contents of `p` in different halves of `my_uint64` (and zero the other half in the latter, but that isn't what's causing a problem). If you wanted to do something *really* disgusting you could implement `getAddressOfPointer` as `return (T **)&Ptr + 1`, I suppose. But ideally we wouldn't be doing anything as highly dodgy when it comes to strict aliasing as that.

Reply:
Right, so the existence of a function that returns the address of internal storage is obviously always going to make abstraction very difficult.  The perfectly portable way to write this is to use a union and a separate bool, and the whole point of this type is that that's a very inefficient way to store this information; we want it to be more compact, and we're willing to accept some non-portable assumptions to get that.  So that's why I think the key question is whether we can change the clients to remove the need for `getAddressOfPointer` first.  Once we do that, the rest of this becomes significantly more trivial.

Reply:
If not, you probably need something like:

```diff
diff --git a/clang/include/clang/AST/ExternalASTSource.h b/clang/include/clang/AST/ExternalASTSource.h
index 385c32edbae0..3f0c63d56d67 100644
--- a/clang/include/clang/AST/ExternalASTSource.h
+++ b/clang/include/clang/AST/ExternalASTSource.h
@@ -25,6 +25,7 @@
 #include "llvm/ADT/SmallVector.h"
 #include "llvm/ADT/iterator.h"
 #include "llvm/Support/PointerLikeTypeTraits.h"
+#include <algorithm>
 #include <cassert>
 #include <cstddef>
 #include <cstdint>
@@ -326,29 +327,43 @@ struct LazyOffsetPtr {
   ///
   /// If the low bit is clear, a pointer to the AST node. If the low
   /// bit is set, the upper 63 bits are the offset.
-  mutable uint64_t Ptr = 0;
+  static constexpr size_t DataSize = std::max(sizeof(uint64_t), sizeof(T *));
+  alignas(uint64_t) alignas(T *) mutable char Data[DataSize] = {};
+
+  template <typename U> U &As() const {
+    if (llvm::sys::IsBigEndianHost)
+      return *reinterpret_cast<U *>(Data + DataSize - sizeof(U));
+    else
+      return *reinterpret_cast<U *>(Data);
+  }
+
+  char &AsLSB() const { return As<char>(); }
+  T *&AsPtr() const { return As<T *>(); }
+  uint64_t &AsU64() const { return As<uint64_t>(); }
 
 public:
   LazyOffsetPtr() = default;
-  explicit LazyOffsetPtr(T *Ptr) : Ptr(reinterpret_cast<uint64_t>(Ptr)) {}
+  explicit LazyOffsetPtr(T *Ptr) : Data() { AsPtr() = Ptr; }
 
-  explicit LazyOffsetPtr(uint64_t Offset) : Ptr((Offset << 1) | 0x01) {
+  explicit LazyOffsetPtr(uint64_t Offset) : Data() {
     assert((Offset << 1 >> 1) == Offset && "Offsets must require < 63 bits");
     if (Offset == 0)
-      Ptr = 0;
+      AsPtr() = NULL;
+    else
+      AsU64() = (Offset << 1) | 0x01;
   }
 
   LazyOffsetPtr &operator=(T *Ptr) {
-    this->Ptr = reinterpret_cast<uint64_t>(Ptr);
+    AsPtr() = Ptr;
     return *this;
   }
 
   LazyOffsetPtr &operator=(uint64_t Offset) {
     assert((Offset << 1 >> 1) == Offset && "Offsets must require < 63 bits");
     if (Offset == 0)
-      Ptr = 0;
+      AsPtr() = NULL;
     else
-      Ptr = (Offset << 1) | 0x01;
+      AsU64() = (Offset << 1) | 0x01;
 
     return *this;
   }
@@ -356,15 +371,15 @@ public:
   /// Whether this pointer is non-NULL.
   ///
   /// This operation does not require the AST node to be deserialized.
-  explicit operator bool() const { return Ptr != 0; }
+  explicit operator bool() const { return isOffset() || AsPtr() != NULL; }
 
   /// Whether this pointer is non-NULL.
   ///
   /// This operation does not require the AST node to be deserialized.
-  bool isValid() const { return Ptr != 0; }
+  bool isValid() const { return isOffset() || AsPtr() != NULL; }
 
   /// Whether this pointer is currently stored as an offset.
-  bool isOffset() const { return Ptr & 0x01; }
+  bool isOffset() const { return AsLSB() & 0x01; }
 
   /// Retrieve the pointer to the AST node that this lazy pointer points to.
   ///
@@ -375,9 +390,9 @@ public:
     if (isOffset()) {
       assert(Source &&
              "Cannot deserialize a lazy pointer without an AST source");
-      Ptr = reinterpret_cast<uint64_t>((Source->*Get)(OffsT(Ptr >> 1)));
+      AsPtr() = (Source->*Get)(OffsT(AsU64() >> 1));
     }
-    return reinterpret_cast<T*>(Ptr);
+    return AsPtr();
   }
 
   /// Retrieve the address of the AST node pointer. Deserializes the pointee if
@@ -385,7 +400,7 @@ public:
   T **getAddressOfPointer(ExternalASTSource *Source) const {
     // Ensure the integer is in pointer form.
     (void)get(Source);
-    return reinterpret_cast<T**>(&Ptr);
+    return &AsPtr();
   }
 };
 
```

Reply:
>> so we need to preserve at least 64 bits in this structure.

> Hmm. We're only really preserving 63 bits here, so that might be a problem.

Sorry for misleading. The current decoding for `GlobalDeclID` (64 bits) is to split it as {ModuleFileIndex : 32, DeclIndex : 32}. And both `ModuleFileIndex` and `DeclIndex` starts from 0 to higher values. So it is safe to remove the highest bit of (ModuleFileIndex) by paying the price that upper bound of the number of modules in a single compilation to be only 2^31, which is a reasonable number.